jax/_src/scipy/optimize/lbfgs.py

```diff
@@ -0,0 +1,229 @@
+    nfev: integer total number of objective evaluations performed.
+    ngev: integer total number of jacobian evaluations
+    x_k: array containing the last argument value found during the search. If
+      the search converged, then this value is the argmin of the objective
+      function.
+    f_k: array containing the value of the objective function at `x_k`. If the
+      search converged, then this is the (local) minimum of the objective
+      function.
+    g_k: array containing the gradient of the objective function at `x_k`. If
+      the search converged the l2-norm of this tensor should be below the
+      tolerance.
+    status: integer describing the status:
+      0 = nominal  ,  1 = max iters reached  ,  2 = max fun evals reached
+      3 = max grad evals reached  ,  4 = insufficient progress (ftol)
+      5 = line search failed
+    ls_status: integer describing the end status of the last line search
+  """
+  converged: Union[bool, jnp.ndarray]
+  failed: Union[bool, jnp.ndarray]
+  k: Union[int, jnp.ndarray]
+  nfev: Union[int, jnp.ndarray]
+  ngev: Union[int, jnp.ndarray]
+  x_k: jnp.ndarray
+  f_k: jnp.ndarray
+  g_k: jnp.ndarray
+  s_history: jnp.ndarray
+  y_history: jnp.ndarray
+  rho_history: jnp.ndarray
+  gamma: Union[float, jnp.ndarray]
+  status: Union[int, jnp.ndarray]
+  ls_status: Union[int, jnp.ndarray]
+
+
+def minimize_lbfgs(
+    fun: Callable,
+    x0: jnp.ndarray,
+    maxiter: Optional[int] = None,
+    norm=jnp.inf,
+    maxcor: int = 10,
+    ftol: float = 2.220446049250313e-09,
+    gtol: float = 1e-05,
+    maxfun: Optional[int] = None,
+    maxgrad: Optional[int] = None,
+    maxls: int = 20,
+):
+  """
+  Minimize a function using L-BFGS
+
+  Implements the L-BFGS algorithm from
+    Algorithm 7.5 from Wright and Nocedal, 'Numerical Optimization', 1999, pg. 176-185
+  And generalizes to complex variables from
+     Sorber, L., Barel, M.V. and Lathauwer, L.D., 2012.
+     "Unconstrained optimization of real functions in complex variables"
+     SIAM Journal on Optimization, 22(3), pp.879-898.
+
+  Args:
+    fun: function of the form f(x) where x is a flat ndarray and returns a real scalar.
+      The function should be composed of operations with vjp defined.
+    x0: initial guess
+    maxiter: maximum number of iterations
+    norm: order of norm for convergence check. Default inf.
+    maxcor: maximum number of metric corrections ("history size")
+    ftol: terminates the minimization when `(f_k - f_{k+1}) < ftol`
+    gtol: terminates the minimization when `|g_k|_norm < gtol`
+    maxfun: maximum number of function evaluations
+    maxgrad: maximum number of gradient evaluations
+    maxls: maximum number of line search steps (per iteration)
+
+  Returns:
+    Optimization results.
+  """
+  d = len(x0)
+  dtype = jnp.dtype(x0)
+
+  # ensure there is at least one termination condition
+  if (maxiter is None) and (maxfun is None) and (maxgrad is None):
```

Comment:
nit: you can drop parentheses here: `is` binds more closely than `and`

Reply:
I prefer the parantheses for readability, but thanks for the pointer